Signal/src/ViewControllers/MessageComposeTableViewController.m

```diff
@@ -667,46 +675,49 @@ - (Contact *)contactForIndexPath:(NSIndexPath *)indexPath {
+        if ([_blockedPhoneNumbers containsObject:phoneNumber.toE164]) {
+            return YES;
+        }
+    }
+
+    return NO;
 }
 
-- (void)refreshContacts {
-    [[ContactsUpdater sharedUpdater] updateSignalContactIntersectionWithABContacts:self.contactsManager.allContacts
-                                                                           success:^{
-                                                                               [self updateContacts];
-                                                                           }
-                                                                           failure:^(NSError *error) {
-                                                                               dispatch_async(dispatch_get_main_queue(), ^{
-                                                                                   UIAlertView *alert =
-                                                                                   [[UIAlertView alloc] initWithTitle:NSLocalizedString(@"ERROR_WAS_DETECTED_TITLE", @"")
-                                                                                                              message:NSLocalizedString(@"TIMEOUT_CONTACTS_DETAIL", @"")
-                                                                                                             delegate:nil
-                                                                                                    cancelButtonTitle:NSLocalizedString(@"OK", @"")
-                                                                                                    otherButtonTitles:nil];
-                                                                                   [alert show];
-                                                                                   [self updateAfterRefreshTry];
-                                                                               });
-                                                                           }];
-    
-    if ([self.contacts count] == 0) {
-        [self showLoadingBackgroundView:YES];
+- (BOOL)isCurrentUserContact:(Contact *)contact
+{
+    for (PhoneNumber *phoneNumber in contact.parsedPhoneNumbers) {
+        if ([[phoneNumber toE164] isEqualToString:[TSAccountManager localNumber]]) {
+            return YES;
+        }
     }
+
+    return NO;
 }
 
-- (void)updateContacts {
-    self.contacts = self.contactsManager.signalContacts;
-    dispatch_async(dispatch_get_main_queue(), ^{
-        [self updateSearchResultsForSearchController:self.searchController];
-        [self.tableView reloadData];
-        [self updateAfterRefreshTry];
-    });
+- (NSArray<Contact *> *_Nonnull)filteredContacts
+{
+    NSMutableArray<Contact *> *result = [NSMutableArray new];
+    for (Contact *contact in self.contactsManager.signalContacts) {
+        if (![self isContactBlocked:contact] && ![self isCurrentUserContact:contact]) {
```

Comment:
### 1. messaging self

It looks like you're trying to exclude the current user from the compose view. In my experience, approximately 1/2 of all people think the app is broken when they can't message themselves. I'm hesitant to change it.

### 2. blocked users

It looks like you're excluding blocked contacts from the compose list. On the one hand - I get it. Alice has stated she doesn't want Bob messaging her, she probably doesn't want to message Bob either.

But I'm afraid that when we hide information like this, a non-insignificant number of people will get confused. Maybe Alice accidentally blocked Bob, or forgot she blocked Bob. Why isn't he in her contacts? 

Alice know's if she doesn't want to message Bob. We don't need to enforce that by hiding him here. Instead I think we should show blocked contacts in the compose view like everyone else, then when Alice actually tries to send something to him (clearly she *wants* to communicate at this point) we prompt her with the "unblock to send" action sheet.

The plus side I see to what you've proposed is that maybe Alice doesn't want to have to *see* Bob in her list of contacts at all, but I still think the "Why isn't Bob in my compose list?" is more likely to feel broken.

A compromise might be relegating blocked contacts in a separate section at the end of the list, or appearing in search in a future release.

What do you think @charlesmchen?